### PR TITLE
review and merge zahary's work

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

**Comment by zah** [comment]
I made this pull request a bit earlier, just to verify that there won't be any Github surprises with reusing the same branch name. More commits are coming.

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -566,20 +596,20 @@ proc genWhileStmt(p: PProc, n: PNode) =
   p.blocks[length].id = -p.unique
   p.blocks[length].isLoop = true
   let labl = p.unique.rope
-  addf(p.body, "L$1: while (true) {$n" | "while (true) {$n", [labl])
-  gen(p, n.sons[0], cond)
-  addf(p.body, "if (!$1) break L$2;$n" | "if (!$1) goto L$2;$n",
+  lineF(p, "L$1: while (true) {$n" | "while (true) {$n", [labl])
+  p.nested: gen(p, n.sons[0], cond)
```
Comment (Araq) [inline compiler/jsgen.nim]: Why does this now depend on ``p.nested``?

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -566,20 +596,20 @@ proc genWhileStmt(p: PProc, n: PNode) =
...
+  lineF(p, "if (!$1) break L$2;$n" | "if (!$1) goto L$2;$n",
        [cond.res, labl])
-  genStmt(p, n.sons[1])
-  addf(p.body, "}$n" | "}L$#:;$n", [labl])
+  p.nested: genStmt(p, n.sons[1])
```
Comment (Araq) [inline compiler/jsgen.nim]: Why does this now depend on ``p.nested``?

File: tests/testament/tester.nim [hunk 3]
```diff
@@ -212,20 +212,31 @@ proc cmpMsgs(r: var TResults, expected, given: TSpec, test: TTest) =
 proc generatedFile(path, name: string, target: TTarget): string =
   let ext = targetToExt[target]
   result = path / "nimcache" /
-    (if target == targetJS: path.splitPath.tail & "_" else: "compiler_") &
+    (if target == targetJS: "" else: "compiler_") &
```
Comment (Araq) [inline tests/testament/tester.nim]: Why was this changed?

Comment (zah) [inline tests/testament/tester.nim]: `generatedFile` is used only when the specs needs to perform a check on the generated code. There wasn't a JS test using this until now and the code was just not correct.
